client/my-sites/themes/controller.jsx

```diff
@@ -114,11 +114,17 @@ export function redirectToThemeDetails( { res, params: { site, theme, section }
 	res.redirect( '/theme/' + [ theme, redirectedSection, site ].filter( Boolean ).join( '/' ) );
 }
 
-export function redirectTiers( { res, originalUrl }, next ) {
-	const redirectUrl = originalUrl.replace( /\/(free|premium|type)/g, '' );
-	if ( redirectUrl === originalUrl ) {
+export function redirectTiers( { res, originalUrl, params: { tier } }, next ) {
+	if ( tier === undefined ) {
 		return next();
 	}
 
+	const typeTierRegex = new RegExp( `/type/${ tier }$`, 'g' );
+	const inlineOrPostfixTierRegex = new RegExp( `(?<=/)${ tier }/|/${ tier }$`, 'g' );
```

Comment:
just a little nit: the second regexp can be shortened to `(?<=/)${ tier }(/|$)`, matching both the infix and suffix version at once. Also, neither regexp should need the `g` flag, as there's always at most one match.

Reply:
👋 If the second regexp is shortened then urls like `en/themes/premium` would be rewritten to `en/themes/` instead of `en/themes`. I've added the extra condition to catch the trailing slash.

I will make a PR for removing the `g` flag.

Reply:
https://github.com/Automattic/wp-calypso/pull/59676

Reply:
You're right, I didn't notice that the `(?<=/)` part is not common for both variants.